sdks/go/pkg/beam/core/runtime/exec/input.go

```diff
@@ -0,0 +1,190 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/funcx"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/log"
+)
+
+//go:generate specialize --input=inputs.tmpl --x=data,universals --y=data,universals
+
+// TODO(herohde) 4/26/2017: SideInput representation? We want it to be amenable
+// to the State API. For now, just use Stream.
+
+// TODO(BEAM-3303): need to include windowing as well when resetting values.
+
+// ReusableInput is a resettable value, notably used to unwind iterators cheaply
+// and cache materialized side input across invocations.
+type ReusableInput interface {
+	// Init initializes the value before use.
+	Init() error
+	// Value returns the side input value.
+	Value() reflect.Value
+	// Reset resets the value after use.
+	Reset() error
+}
+
+var (
+	inputs   = make(map[string]func(ReStream) ReusableInput)
+	inputsMu sync.Mutex
+)
+
+// RegisterInput registers an input handler for the given type, such as "func(*int)bool". If
+// multiple input handlers are registered for the same type, the last registration wins.
+func RegisterInput(t reflect.Type, maker func(ReStream) ReusableInput) {
+	inputsMu.Lock()
+	defer inputsMu.Unlock()
+
+	key := t.String()
+	if _, exists := inputs[key]; exists {
+		log.Warnf(context.Background(), "Input for %v already registered. Overwriting.", key)
+	}
+	inputs[key] = maker
+}
+
+type reIterValue struct {
```

Comment:
This casing seems incorrect. Camel-casing is general used to mark word boundaries, but here, it's emphasizing the prefix 're'. I think this should be 'reiterValue'

Reply:
The name is also meant to match the "ReIter" constant. If both should be "Reiter" then we should probably come up with a better name altogether. I prefer to address such a change in a separate PR.

Reply:
I don't have a great suggestion here. I agree this comment is not a blocker for this PR.